compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -2917,91 +2933,534 @@ setRootConfigImpl(mlir::FunctionOpInterface entryPointFn, Operation *op,
+      int64_t globalDimIdx = dimTracker.getGlobalDimIdx(op, pos);
+      if (size <= 0 || !llvm::is_contained(rootOpGlobalDims, globalDimIdx)) {
+        continue;
+      }
+      globalTileSizes[level][globalDimIdx] =
+          updater(globalTileSizes[level][globalDimIdx], size);
+    }
+  };
+  auto align = [](int64_t oldSize, int64_t newSize) {
+    return llvm::alignTo(oldSize, newSize);
+  };
+  auto overwrite = [](int64_t oldSize, int64_t newSize) { return newSize; };
+  // Adjust root op tiling sizes with non-root op.
+  for (auto &[op, vecTileSize] : nonRootOpVecTileSizes) {
+    if (isa<linalg::PackOp>(op)) {
+      // For pack op, align the distribution tile size and overwrite the
+      // vector parallel tile size.
+      adjust(op, vecTileSize, TilingLevel::DistributionTiles, align);
+      adjust(op, vecTileSize, TilingLevel::VectorCommonParallelTiles,
+             overwrite);
+    } else if (auto unpackOp = dyn_cast<linalg::UnPackOp>(op)) {
+      // For unpack op, just overwrite the vector parallel tile size.
+      // However, dimension tracking is expected be broken in the case of
+      // `generic -> unpack`, since only the unpacked dimensions are propated.
+      // To correct this, use the generic op result indexing map to update the
+      // tracking.
+      //
+      // Example: If the generic op result has an affine map
+      //          (d0, d1, d2, d3) -> (d0, d1, d2, d3),
+      //          we use `vecTileSize` to adjust (d2, d3) instead of (d0, d1).
+      auto linalgOp = unpackOp.getSource().getDefiningOp<linalg::LinalgOp>();
+      if (!linalgOp) {
+        continue;
+      }
+      AffineMap indexingMap = linalgOp.getIndexingMapMatchingResult(
+          cast<OpResult>(unpackOp.getSource()));
+      SmallVector<int64_t> adjustedTileSize(linalgOp.getNumLoops(), 0);
+      for (auto [expr, tileSize] : llvm::zip_equal(
+               indexingMap.getResults().take_back(vecTileSize.size()),
+               vecTileSize)) {
+        auto dimExpr = dyn_cast<AffineDimExpr>(expr);
+        if (!dimExpr) {
+          continue;
+        }
+        adjustedTileSize[dimExpr.getPosition()] = tileSize;
+      }
+      adjust(linalgOp.getOperation(), adjustedTileSize,
+             TilingLevel::VectorCommonParallelTiles, overwrite);
+    }
   }
-  for (auto [pos, size] : llvm::enumerate(vecTileSizes)) {
-    if (!size)
+
+  // Adjust root op tiling sizes with root result element type.
+  // Make sure the innermost tile size times element size is multiple
+  // of byte bits. This is required for now because we do not fully
+  // support sub-byte vector stores. Once vector stores are supported
+  // then this can be eliminated. Note that emulating sub-byte sized vector
+  // loads and stores will have a performance impact.
+  auto resultTypes = rootOperation->getResultTypes();
+  if (!resultTypes.empty()) {
+    Type elementType = cast<ShapedType>(resultTypes[0]).getElementType();
```

Comment:
optional nit: I think you can use `getElementTypeOrSelf`. It is reasonable to me because linalg ops take either scalar operands or shaped type operands.